STN-54: Main Navigation

kaseybon opened the pull request:
# READY FOR REVIEW

## Summary
This pull request adds behavior changes to the main navigation and page layout:
- The sidebar navigation is hidden at small-medium screen sizes but visible at the large breakpoint
- the mobile nav will include all pages available at small-medium screens
- the desktop nav will only include the top 2 levels of pages available at large screens

## Need Review By (Date)
asap

## Urgency
high

## Steps to Test
1. Build both CSS and JS
2. Run `npm test` to confirm there are no errors
3. Adjust the main navigation to show 5 levels instead of the default 2:
    `Structure > Blocks Layout` and then configure the Main Navigation block in the `menu` region.
4. Confirm that all the navigation meets all acceptance criteria (link in the Jira card)

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)

kaseybon committed:
feat: hide the secondary nav until large screens

kaseybon committed:
feat: JS and HTML refactor to the main navigation

- since we want the mobile navigation to behave differently than the Decanter nav does we needed to write custom JS
- these changes expect there to be a secondary nav on the page that is hidden at small-medium screen sizes but visible at large screens
- the mobile nav will include all pages available at small-medium screens
- the desktop nav will only include the top 2 levels of pages availble at large screens

STN-54

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/objects/_layouts.three_column_w_image.scss:
```diff
@@ -8,6 +8,10 @@
 
   &.hb-page-width {
     @include hb-page-width;
+    // scss-lint:disable ImportantRule
```
I've added this in my PR, so I think you can remove all the new green stuff :) it will all be part of the hb-page-width mixin now

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/objects/_layouts.three_column.scss:
```diff
@@ -8,27 +8,29 @@
 
   &.hb-page-width {
     @include hb-page-width;
+    // scss-lint:disable ImportantRule
```
I've added this in my PR, so I think you can remove all the new green stuff :) it will all be part of the hb-page-width mixin now

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_main-nav.scss:
```diff
@@ -26,5 +16,89 @@
+    margin-top: 8px;
+    margin-bottom: 8px;
+
+    &[aria-hidden="true"] {
```
do we need some noJS fallbacks for aria-hidden classes that may or may not happen if JS is turned off?

silverli replied:
you can use the .no-js class that will be on the top html element too

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/nested-toggler.js:
```diff
@@ -0,0 +1,72 @@
+    let isHidden;
+
+    const collapseMenu = () => {
+      togglerContent.setAttribute('aria-hidden', true);
```
I am on the fence with the aria-hidden stuff. technically, I think you should never hide a tabbable element (such as the a tags under the ul). I think it would be best to apply display: none; instead. we could add a class and then do that in the styles

silverli replied:
I think what happens is that even if you set aria-hidden to true, you can still tab around, but nothing will be said, which is not a great experience, and can be very confusing for screen reader users

silverli replied:
oh i see.. you do set it to display:none;  still feels like you could add a class instead of aria-hidden since it's not doing much  when the display CSS is used

kaseybon replied:
I prefer to use aria-hidden and then target that attribute with the CSS because it helps prevent against future changes breaking things. If we rely on a class and something happens where the class is being toggled but aria attributes are not we probably won't notice right away.

silverli commented on docroot/themes/humsci/humsci_basic/templates/menus/macros/nav-menu.twig:
```diff
@@ -1,6 +1,6 @@
 {#
 /**
- * Macro for creating nested menus.
+ * Macro for creating nested menus in the main navigation.
```
in `docroot/themes/humsci/humsci_basic/templates/menus/menu--main.html.twig` can you remove the 'no-js' class that is getting added?

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/menu-toggle.js:
```diff
@@ -0,0 +1,56 @@
+const collapseNav = () => {
+  menuToggle.setAttribute('aria-expanded', false);
+  mainMenu.setAttribute('aria-hidden', true);
+  menuToggle.innerHTML = "Menu";
```
we actually dont have to set the text at all here. If we keep it "Menu" in the menu block code template, then that content can be translated by drupal if we ever need. Also, a screenreader user will hear, with the aria-expanded attribute: Menu Button Expanded / Menu button collapsed and that's all they really need to know!

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/menu-toggle.js:
```diff
@@ -0,0 +1,56 @@
+  isHidden = true;
+}
+
+const showNav = () => {
```
I tried out a DRY-er function for these show/CollapseNav functions, as well as the ones in nested-toggler:

if you want to make a new file called changeNav.js in this same main-nav directory, add the following:


```
const changeNav = (toggle, parent, isOpen) => {
  toggle.setAttribute('aria-expanded', isOpen);
  parent.setAttribute('aria-hidden', !isOpen);
// just set isHidden variable in the local file underneath the call to this function
}

export default changeNav;

```

and then to import it and use it in **this file** you can do:

```
import changeNav from './changeNav';


// rather than calling showNav or collapseNav, you can just do:

 // Toggle the nav when the the button is clicked
  menuToggle.addEventListener('click', () => {
    if (isHidden) {
      changeNav(menuToggle, mainMenu, true);
      isHidden = false;
    } else {
      changeNav(menuToggle, mainMenu, false);
      isHidden = true;
    }
  });
```

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/menu-toggle.js:
```diff
@@ -0,0 +1,56 @@
+
+if (menuToggle) {
+  // If below a certain breakpoint, hide the nav on page load
+  if (window.innerWidth < mobileNavBreakpoint) {
```
I think this should already be set up by default in the template code and styles. I don't think we should use JS to accomplish it. I would use no-JS and .js styles to do so instead in the styles.

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/menu-toggle.js:
```diff
@@ -0,0 +1,56 @@
+}
+
+if (menuToggle) {
+  // If below a certain breakpoint, hide the nav on page load
```
this is more like if there is JS on the page, then collapse the nav. we should set aria-expanded to true on the button trigger in this template code `docroot/themes/humsci/humsci_basic/templates/menus/menu--main.html.twig` by default on page load, and then this JS can come in if there is JS enabled on the page

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/menu-toggle.js:
```diff
@@ -0,0 +1,56 @@
+  window.addEventListener('resize', () => {
+    windowWidth = window.innerWidth;
+
+    // When resizing from mobile to desktop, show the navigation
```
can you add this comment here: 
```
// When resizing from mobile to desktop, ensure navigation is displayed, not hidden
// If wasDesktopSize is false, it means we haven't gotten there yet and will to run this check
// Otherwise, if wasDesktopSize is true, we are above the mobileNavBreakpoint and don't need to keep showingNav
```

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/menu-toggle.js:
```diff
@@ -0,0 +1,56 @@
+    windowWidth = window.innerWidth;
+
+    // When resizing from mobile to desktop, show the navigation
+    if (windowWidth >= mobileNavBreakpoint) {
```
can you then amend the conditional check to the following so showNav doesnt get continuously run over and over as we resize the screen upwards?

`if (windowWidth >= mobileNavBreakpoint && !wasDesktopSize)`

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/menu-toggle.js:
```diff
@@ -0,0 +1,56 @@
+    if (windowWidth < mobileNavBreakpoint && wasDesktopSize) {
+      collapseNav();
+
+      // This keeps the navigation from collapsing every time the screen is resized
```
can you add a comment like this one, or saying something similiar:

`// After the first time we resize to below the mobileNavBreakpoint, reset wasDesktopSize var`

silverli commented on docroot/themes/humsci/humsci_basic/templates/menus/macros/nav-menu.twig:
```diff
@@ -27,9 +27,26 @@
-      <a{{ link_attributes }}>{{ item.title }}</a>
-      {% if item.below %}
-        {{ menus.nav_menu(item.below, menu_level + 1, class_prefix) }}
+      {% if item.url.toString() is empty %}
```
can you explain this if check a bit more? I am not exactly sure how or why we decide between button or link

silverli replied:
or adding a comment would be awesome

silverli commented on docroot/themes/humsci/humsci_basic/templates/menus/macros/nav-menu.twig:
```diff
@@ -27,9 +27,26 @@
-        {{ menus.nav_menu(item.below, menu_level + 1, class_prefix) }}
+      {% if item.url.toString() is empty %}
+        {% set link_attributes = link_attributes.addClass('hb-nested-toggler') %}
+        <a{{ link_attributes }} id="{{ item.title|replace({' ': ''}) }}{{ menu_level }}">{{ item.title }}</a>
```
should we add role=button?

kaseybon replied:
I added the role and also adjusted the JS to make sure that the spacebar will also toggle the anchor tags that are buttons.

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/nested-toggler.js:
```diff
@@ -0,0 +1,72 @@
+    const toggler = togglers[i];
+    const togglerID = toggler.getAttribute('id');
+    const togglerContent = document.querySelector(`[aria-labelledby="${togglerID}"]`);
+    const togglerParent = toggler.closest('.su-main-nav__item--parent');
```
.closest isnt supported in ie11

silverli replied:
.parentNode might work though!

kaseybon replied:
We have Babel to handle this. It looks like there is an error in the secondary nav JS from Decanter that is causing our JS to not run. ~I'm going to move all Decanter imports to the bottom of the index.js for now so our JS still runs.~ That didn't do anything.

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/nested-toggler.js:
```diff
@@ -0,0 +1,72 @@
+    const togglerID = toggler.getAttribute('id');
+    const togglerContent = document.querySelector(`[aria-labelledby="${togglerID}"]`);
+    const togglerParent = toggler.closest('.su-main-nav__item--parent');
+    const shouldBeOpen = togglerParent.classList.contains('su-main-nav__item--active-trail') ? true : false;
```
i find this naming a bit confusing

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/nested-toggler.js:
```diff
@@ -0,0 +1,72 @@
+let windowWidth = window.innerWidth;
+
+if (togglers) {
+  for (let i = 0; i < togglers.length; i += 1) {
```
hmm I dont see this logic happening to all my nested-togglers. Only the first one it seems 🤔 maybe we can pair on this

kaseybon replied:
I updated the description with one config change that I had forgotten about. Let me know if this fixes things.

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/nested-toggler.js:
```diff
@@ -0,0 +1,72 @@
+      }
+    });
+
+    // When tabbing through the navigation the previously opened dropdown closes
```
are these two event listeners for desktop only?

kaseybon replied:
Yep. We don't want to close the mobile nested navs in either of these scenarios.

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_main-nav.scss:
```diff
@@ -1,20 +1,10 @@
-// Override some decanter styles for Main Nav
```
I would keep some version of this comment here

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_main-nav.scss:
```diff
@@ -26,5 +16,89 @@
+  }
+
+  &__menu-container {
+    padding-left: 20px;
```
px not rem? just wanted to check

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_main-nav.scss:
```diff
@@ -26,5 +16,89 @@
+    position: absolute;
+    top: 0;
+    right: 0;
+    width: 44px;
```
px or rem?

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_masthead.scss:
```diff
@@ -12,36 +13,15 @@
+    @include grid-media-min('sm') {
+      flex: unset;
+      width: 100% !important;
+      max-width: none;
```
never seen this! interesting

silverli commented on docroot/themes/humsci/humsci_basic/src/js/main-nav/nested-toggler.js:
```diff
@@ -0,0 +1,72 @@
+      windowWidth = window.innerWidth;
+
+      // When resizing from mobile to desktop, show the navigation
+      if (windowWidth >= mobileNavBreakpoint) {
```
youll need a wasDesktopSize here too so collapseMenu doesn't keep getting called. I think the logic would go like this:


```
if (windowWidth >= mobileNavBreakpoint && !wasDesktopSize) {
  collapseMenu();
  wasDesktopSize = true;
}

```

but I wonder if you need this since we'll close the nav on desktop in the menu-toggle.js?

kaseybon replied:
Adding this creates a new edge case when resizing the browser. Personally I think this edge case happening at desktops isn't a huge issue but the new issue it creates is. I would suggest we skip this and if for some reason the edge case is a problem in the future we can tackle it then.

silverli replied:
that is fine with me!! not a huge deal :)

silverli reviewed (changes requested):
This is really great work. I know there are a lot of comments, but I think they are mostly pretty small. We can pair up in the morning too if we want to see if there are some things that can be implemented later.

The one big thing I had a concern with was that for my nested-togglers, it looked like only the first set was getting the appropriate aria attributes, and the others weren't.

I also wanted to check that we should be seeing 5 levels deep on mobile, and 2 on desktop? I am not sure I am seeing this behavior, I might be missing some nav config.

kaseybon committed:
feat: improve nav accessibility

- when an anchor tag is a toggler the spacebar will open/close the dropdown
- added comments in template to explain the markup

STN-54

kaseybon committed:
fix: responding to scss and template feedback

kaseybon committed:
fix: responding to js and template feedback

kaseybon committed:
fix: check for navigation items that are not links but also don't have nested pages

- This is a scenario that shouldn't happen but could, this code prevents the rest of the navigation from breaking

STN-54

silverli commented on docroot/themes/humsci/humsci_basic/templates/menus/macros/nav-menu.twig:
```diff
@@ -11,7 +11,16 @@
     {% set link_attributes = link_attributes.addClass(class_prefix ~ "__link") %}
-    {% set link_attributes = link_attributes.setAttribute('href', item.url|render) %}
+    {#
+      If there isn't a URL we want to set the href to #0 to prevent a page refresh
```
good catch 👏

kaseybon committed:
fix: typo

silverli reviewed (approved):
Good work!
